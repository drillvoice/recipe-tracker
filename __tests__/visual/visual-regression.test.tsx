import { render } from '@testing-library/react';

// Mock all external dependencies for consistent visual testing
jest.mock('@/lib/offline-storage', () => ({
  saveMeal: jest.fn(),
  updateMeal: jest.fn(),
  deleteMeal: jest.fn(),
  getAllMeals: () => Promise.resolve([
    {
      id: '1',
      mealName: 'Spaghetti Carbonara',
      date: {
        toDate: () => new Date('2024-03-15T10:30:00Z'),
        toMillis: () => new Date('2024-03-15T10:30:00Z').getTime(),
      },
      uid: 'test-uid',
      hidden: false,
      tags: ['Italian', 'Dinner']
    },
    {
      id: '2',
      mealName: 'Caesar Salad',
      date: {
        toDate: () => new Date('2024-03-14T18:45:00Z'),
        toMillis: () => new Date('2024-03-14T18:45:00Z').getTime(),
      },
      uid: 'test-uid',
      hidden: false,
      tags: ['Salad']
    },
    {
      id: '3',
      mealName: 'Hidden Meal',
      date: {
        toDate: () => new Date('2024-03-13T12:00:00Z'),
        toMillis: () => new Date('2024-03-13T12:00:00Z').getTime(),
      },
      uid: 'test-uid',
      hidden: true,
      tags: []
    }
  ]),
  hideMealsByName: jest.fn(),
}));

jest.mock('@/lib/firebase', () => ({
  auth: {
    currentUser: { uid: 'test-uid' }
  }
}));

jest.mock('firebase/auth', () => ({
  onAuthStateChanged: (
    _auth: unknown,
    cb: (user: { uid: string; isAnonymous: boolean }) => void
  ) => {
    cb({ uid: 'test-uid', isAnonymous: true });
    return () => {};
  }
}));

// Import components after mocking
const AddPage = require('@/pages/index').default;
const DishesPage = require('@/pages/dishes').default;
const TagsPage = require('@/pages/tags').default;
const AccountPage = require('@/pages/account').default;

/**
 * Visual regression testing utilities
 * 
 * Note: This is a basic setup for visual regression testing.
 * For full implementation, you would typically use tools like:
 * - Chromatic (for Storybook)
 * - Percy (for CI/CD integration)
 * - Playwright with screenshots
 * - React Testing Library with jest-image-snapshot
 */

// Mock dates to ensure consistent snapshots
const mockDate = new Date('2024-03-15T10:30:00Z');
const originalDate = Date;

beforeAll(() => {
  // Mock Date constructor with better compatibility
  global.Date = class extends originalDate {
    constructor(...args: ConstructorParameters<typeof Date>) {
      if (args.length) {
        super(...args);
      } else {
        super(mockDate);
      }
    }

    static now() {
      return mockDate.getTime();
    }

    getTime() {
      if (this.valueOf() === mockDate.valueOf()) {
        return mockDate.getTime();
      }
      return super.getTime();
    }

    toISOString() {
      if (this.valueOf() === mockDate.valueOf()) {
        return mockDate.toISOString();
      }
      return super.toISOString();
      }
  } as unknown as DateConstructor;
});

afterAll(() => {
  global.Date = originalDate;
});

describe.skip('Visual Regression Tests', () => {
  describe('Page Layouts', () => {
    test('Add Meal page renders consistently', () => {
      const { container } = render(<AddPage />);
      
      // Basic structure checks
      expect(container.querySelector('.container')).toBeInTheDocument();
      expect(container.querySelector('.top-nav')).toBeInTheDocument();
      expect(container.querySelector('.form')).toBeInTheDocument();
      expect(container.querySelector('.version-indicator')).toBeInTheDocument();
      
      // Check for key elements that affect layout
      expect(container.querySelector('input[placeholder="Enter meal name..."]')).toBeInTheDocument();
      expect(container.querySelector('input[type="date"]')).toBeInTheDocument();
      expect(container.querySelector('button')).toBeInTheDocument();
      
      // Take a snapshot for future comparison
      expect(container.firstChild).toMatchSnapshot('add-meal-page');
    });

    test('Tags page renders consistently', async () => {
      const { container } = render(<TagsPage />);

      // Wait for async data loading
      await new Promise(resolve => setTimeout(resolve, 100));

      // Basic structure checks
      expect(container.querySelector('.container')).toBeInTheDocument();
      expect(container.querySelector('.top-nav')).toBeInTheDocument();

      // Should have placeholder content
      await new Promise(resolve => setTimeout(resolve, 100));

      expect(container.firstChild).toMatchSnapshot('tags-page');
    });

    test('Ideas page renders consistently', async () => {
      const { container } = render(<DishesPage />);
      
      // Wait for async data loading
      await new Promise(resolve => setTimeout(resolve, 100));
      
      // Basic structure checks
      expect(container.querySelector('.container')).toBeInTheDocument();
      expect(container.querySelector('.page-header')).toBeInTheDocument();
      
      expect(container.firstChild).toMatchSnapshot('ideas-page');
    });

    test('Account page renders consistently', () => {
      const { container } = render(<AccountPage />);

      // Basic structure checks
      expect(container.querySelector('.container')).toBeInTheDocument();
      expect(container.querySelector('.data-section')).toBeInTheDocument();

      expect(container.firstChild).toMatchSnapshot('account-page');
    });
  });

  describe('Component States', () => {
    test('Empty state renders consistently', async () => {
      // Override mock to return empty data
      const mockGetAllMealsEmpty = jest.fn().mockResolvedValue([]);
      jest.doMock('@/lib/offline-storage', () => ({
        getAllMeals: mockGetAllMealsEmpty,
        saveMeal: jest.fn(),
        updateMeal: jest.fn(),
        deleteMeal: jest.fn(),
        hideMealsByName: jest.fn(),
      }));
      
      const { container } = render(<TagsPage />);

      // Wait for loading to complete
      await new Promise(resolve => setTimeout(resolve, 100));

      expect(container.firstChild).toMatchSnapshot('tags-empty-state');
    });

    test('Loading state renders consistently', () => {
      // Create a component that shows loading state
      const LoadingComponent = () => (
        <main className="container">
          <div className="top-nav">
            <a className="nav-item" href="/">+ Add</a>
            <a className="nav-item active" href="/history">History</a>
            <a className="nav-item" href="/dishes">Dishes</a>
            <a className="nav-item" href="/account">Account</a>
          </div>
          <h1>History</h1>
          <p>Loading meals...</p>
        </main>
      );
      
      const { container } = render(<LoadingComponent />);
      expect(container.firstChild).toMatchSnapshot('loading-state');
    });

    test('Error state renders consistently', () => {
      // Create a component that shows error state
      const ErrorComponent = () => (
        <main className="container">
          <div className="top-nav">
            <a className="nav-item" href="/">+ Add</a>
            <a className="nav-item active" href="/history">History</a>
            <a className="nav-item" href="/dishes">Dishes</a>
            <a className="nav-item" href="/account">Account</a>
          </div>
          <h1>History</h1>
          <p className="error-message">Error loading meals: Network error</p>
        </main>
      );
      
      const { container } = render(<ErrorComponent />);
      expect(container.firstChild).toMatchSnapshot('error-state');
    });
  });

  describe('Responsive Layouts', () => {
    test('Mobile layout renders consistently', () => {
      // Mock mobile viewport
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 375, // iPhone width
      });
      
      const { container } = render(<AddPage />);
      
      // Check mobile-specific elements
      expect(container.querySelector('.container')).toBeInTheDocument();
      
      expect(container.firstChild).toMatchSnapshot('add-meal-mobile');
    });

    test('Tablet layout renders consistently', () => {
      // Mock tablet viewport
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 768, // iPad width
      });
      
      const { container } = render(<AddPage />);
      expect(container.firstChild).toMatchSnapshot('add-meal-tablet');
    });

    test('Desktop layout renders consistently', () => {
      // Mock desktop viewport
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 1200,
      });
      
      const { container } = render(<AddPage />);
      expect(container.firstChild).toMatchSnapshot('add-meal-desktop');
    });
  });

  describe('Interactive States', () => {
    test('Dialog modal renders consistently', () => {
      // Create a component with modal open
      const ModalComponent = () => (
        <div>
          <main className="container">
            <h1>Test Page</h1>
          </main>
          <div className="dialog-overlay">
            <div className="dialog-content">
              <h2 className="dialog-title">Delete Meal</h2>
              <p className="dialog-message">Are you sure you want to delete "Pizza" from Mar 15?</p>
              <div className="dialog-actions">
                <button className="dialog-btn dialog-btn-secondary">Cancel</button>
                <button className="dialog-btn dialog-btn-danger">Delete</button>
              </div>
            </div>
          </div>
        </div>
      );
      
      const { container } = render(<ModalComponent />);
      expect(container.firstChild).toMatchSnapshot('modal-dialog');
    });

    test('Form validation states render consistently', () => {
      const FormWithValidation = () => (
        <main className="container">
          <div className="form">
            <label>
              Meal Name
              <input
                defaultValue="Test Meal"
                className="edit-input"
                style={{ borderColor: '#16a34a' }} // Success state
                readOnly
              />
            </label>
            <label>
              Date
              <input
                type="date"
                defaultValue=""
                className="edit-input"
                style={{ borderColor: '#dc2626' }} // Error state
                readOnly
              />
            </label>
          </div>
        </main>
      );

      const { container } = render(<FormWithValidation />);
      expect(container.firstChild).toMatchSnapshot('form-validation-states');
    });
  });
});