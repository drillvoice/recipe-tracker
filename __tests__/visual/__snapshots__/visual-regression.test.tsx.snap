// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visual Regression Tests Component States Empty state renders consistently: history-empty-state 1`] = `
<main
  class="container"
>
  <nav
    class="top-nav"
  >
    <a
      class="nav-item "
      href="/"
    >
      + Add
    </a>
    <a
      class="nav-item active"
      href="/history"
    >
      History
    </a>
    <a
      class="nav-item "
      href="/ideas"
    >
      Ideas
    </a>
    <a
      class="nav-item "
      href="/account"
    >
      Account
    </a>
  </nav>
  <h1>
    History
  </h1>
  <p
    class="error-message"
  >
    Error loading meals: 
    this is not a Date object.
  </p>
</main>
`;

exports[`Visual Regression Tests Component States Error state renders consistently: error-state 1`] = `
<main
  class="container"
>
  <div
    class="top-nav"
  >
    <a
      class="nav-item"
      href="/"
    >
      + Add
    </a>
    <a
      class="nav-item active"
      href="/history"
    >
      History
    </a>
    <a
      class="nav-item"
      href="/ideas"
    >
      Ideas
    </a>
    <a
      class="nav-item"
      href="/account"
    >
      Account
    </a>
  </div>
  <h1>
    History
  </h1>
  <p
    class="error-message"
  >
    Error loading meals: Network error
  </p>
</main>
`;

exports[`Visual Regression Tests Component States Loading state renders consistently: loading-state 1`] = `
<main
  class="container"
>
  <div
    class="top-nav"
  >
    <a
      class="nav-item"
      href="/"
    >
      + Add
    </a>
    <a
      class="nav-item active"
      href="/history"
    >
      History
    </a>
    <a
      class="nav-item"
      href="/ideas"
    >
      Ideas
    </a>
    <a
      class="nav-item"
      href="/account"
    >
      Account
    </a>
  </div>
  <h1>
    History
  </h1>
  <p>
    Loading meals...
  </p>
</main>
`;

exports[`Visual Regression Tests Interactive States Dialog modal renders consistently: modal-dialog 1`] = `
<div>
  <main
    class="container"
  >
    <h1>
      Test Page
    </h1>
  </main>
  <div
    class="dialog-overlay"
  >
    <div
      class="dialog-content"
    >
      <h2
        class="dialog-title"
      >
        Delete Meal
      </h2>
      <p
        class="dialog-message"
      >
        Are you sure you want to delete "Pizza" from Mar 15?
      </p>
      <div
        class="dialog-actions"
      >
        <button
          class="dialog-btn dialog-btn-secondary"
        >
          Cancel
        </button>
        <button
          class="dialog-btn dialog-btn-danger"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Visual Regression Tests Interactive States Form validation states render consistently: form-validation-states 1`] = `
<main
  class="container"
>
  <div
    class="form"
  >
    <label>
      Meal Name
      <input
        class="edit-input"
        style="border-color: #16a34a;"
        value="Test Meal"
      />
    </label>
    <label>
      Date
      <input
        class="edit-input"
        style="border-color: #dc2626;"
        type="date"
        value=""
      />
    </label>
  </div>
</main>
`;

exports[`Visual Regression Tests Page Layouts Account page renders consistently: account-page 1`] = `
<main
  class="container"
>
  <nav
    class="top-nav"
  >
    <a
      class="nav-item "
      href="/"
    >
      + Add
    </a>
    <a
      class="nav-item "
      href="/history"
    >
      History
    </a>
    <a
      class="nav-item "
      href="/ideas"
    >
      Ideas
    </a>
    <a
      class="nav-item active"
      href="/account"
    >
      Account
    </a>
  </nav>
  <h1>
    Account
  </h1>
  <div
    class="auth-card"
  >
    <div
      class="auth-tabs"
    >
      <button
        class="auth-tab active"
      >
        Sign In
      </button>
      <button
        class="auth-tab "
      >
        Sign Up
      </button>
    </div>
    <div
      class="auth-fields"
    >
      <p
        class="subtitle"
      >
        Sign in to sync your meals across devices
      </p>
      <label>
        Email Address
        <input
          placeholder="Enter your email..."
          type="email"
          value=""
        />
      </label>
      <label>
        Password
        <input
          minlength="6"
          placeholder="Enter your password..."
          type="password"
          value=""
        />
      </label>
      <button
        class="auth-submit"
      >
        Sign In
      </button>
      <button
        class="auth-submit secondary"
      >
        Forgot Password?
      </button>
    </div>
  </div>
</main>
`;

exports[`Visual Regression Tests Page Layouts History page renders consistently: history-page 1`] = `
<main
  class="container"
>
  <nav
    class="top-nav"
  >
    <a
      class="nav-item "
      href="/"
    >
      + Add
    </a>
    <a
      class="nav-item active"
      href="/history"
    >
      History
    </a>
    <a
      class="nav-item "
      href="/ideas"
    >
      Ideas
    </a>
    <a
      class="nav-item "
      href="/account"
    >
      Account
    </a>
  </nav>
  <h1>
    History
  </h1>
  <p
    class="error-message"
  >
    Error loading meals: 
    this is not a Date object.
  </p>
</main>
`;
